### PR TITLE
Support variadic parameter in `array_replace`

### PR DESCRIPTION
**Pull request by herndlm**
Closes https://github.com/phpstan/phpstan/issues/7020

It was not my intention to completely refactor this extension, but it kind of happened while adding variadic support, sorry 😅 I normally would try to separate such things in multiple PRs. But the fix needed the refactor, the tests here were good enough and it's all quite local I think. Variadic support is only added via that one line checking `->unpack` CC @staabm

### CODE REVIEW

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php [hunk 1]
```diff
@@ -24,68 +24,57 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
...
-
-			if (!$arrayType->isArray()->yes()) {
-				return null;
-			}
-
+		$nonEmptyArray = false;
+		foreach ($arrayTypes as $arrayType) {
 			$keyTypes[] = $arrayType->getIterableKeyType();
 			$valueTypes[] = $arrayType->getIterableValueType();
+			if ($nonEmptyArray || !$arrayType->isIterableAtLeastOnce()->yes()) {
```
Comment (herndlm) [inline src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php:44]: While I do like early exists, I'm not the biggest fan of this which the code-style kind of forced me too. Let me know if you have any better ideas, I wanted to avoid having to loop over the arrays again just for that. But it would read nicer for sure :)

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php [hunk 2]
```diff
@@ -24,68 +24,57 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
...
+				continue;
+			}
+
+			$nonEmptyArray = true;
```
Comment (herndlm) [inline src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php:48]: While I do like early exists, I'm not the biggest fan of this and the block above, which the code-style kind of forced me to do. Let me know if you have any better ideas, I wanted to avoid having to loop over the arrays again just for that. But it would read nicer for sure :)

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php [hunk 3]
```diff
@@ -24,68 +24,57 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
...
-			if (!$arrayType->isArray()->yes()) {
-				return null;
-			}
-
+		$nonEmptyArray = false;
+		foreach ($arrayTypes as $arrayType) {
 			$keyTypes[] = $arrayType->getIterableKeyType();
 			$valueTypes[] = $arrayType->getIterableValueType();
+			if ($nonEmptyArray || !$arrayType->isIterableAtLeastOnce()->yes()) {
+				continue;
```
Comment (herndlm) [inline src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php]: While I do like early exists, I'm not the biggest fan of this, which the code-style kind of forced me to do. Let me know if you have any better ideas, I wanted to avoid having to loop over the arrays again just for that. But it would read nicer for sure :)

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php [hunk 4]
```diff
@@ -24,68 +24,57 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
 
 	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
 	{
-		$args = $functionCall->getArgs();
+		$arrayTypes = $this->collectArrayTypes($functionCall, $scope);
 
-		if (count($args) < 1) {
+		if (count($arrayTypes) === 0) {
 			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
```
Comment (staabm) [inline src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php]: I think you could return `null` here to simplify it

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php (hunk 3 above)
Comment (staabm) [reply src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php]: I don't like this style either.

I write code like I want it to be and let the auto-cs-fixer decide what todo with it

**Comment by staabm** [comment]
> Variadic support is only added via that one line checking ->unpack

I agree its hard to follow and the fix itself would have been a great separate PR. but lets see how ondrej thinks about that

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php (hunk 3 above)
Comment (herndlm) [reply src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php]: ..which I did and phpcbf adapted it to that, hehe

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php (hunk 4 above)
Comment (herndlm) [reply src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php]: oh right, I forgot. I definitely prefer the new way, thx for the reminder! adapted

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php (hunk 3 above)
Comment (herndlm) [reply src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php]: I kind of "tricked" CS by moving this up, should be simpler now :)

**Comment by ondrejmirtes** [comment]
Thank you.

File: src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php (hunk 3 above)
Comment (ondrejmirtes) [reply src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php]: If you think the CS sniff can be improved, open a specific discussion in https://github.com/slevomat/coding-standard (and feel free to ping me there).
